DirectProgramming/C++SYCL_FPGA/Tutorials/DesignPatterns/explicit_data_movement/src/explicit_data_movement.cpp

```diff
@@ -81,75 +73,98 @@ double SubmitImplicitKernel(queue& q, std::vector<T>& in, std::vector<T>& out,
 // This version of the kernel demonstrates explicit data movement
 // through explicit USM.
 //
-template<typename T>
-double SubmitExplicitKernel(queue& q, std::vector<T>& in,
-                            std::vector<T>& out, size_t size) {
-#if defined (IS_BSP)
-  // allocate the device memory
-  T* in_ptr = malloc_device<T>(size, q);
-  T* out_ptr = malloc_device<T>(size, q);
+template <typename T>
+double SubmitExplicitKernel(sycl::queue &q, std::vector<T> &in,
+                            std::vector<T> &out, size_t size) {
+#if defined(IS_BSP)
+  // USM device allocations are more commonly supported by FPGA boards than
+  // other types of USM allocations like host and shared allocations.
+  // Allocate the device memory
+  T *in_ptr = sycl::malloc_device<T>(
+      size, q,
+      sycl::ext::intel::experimental::property::usm::buffer_location(0));
+  T *out_ptr = sycl::malloc_device<T>(
+      size, q,
+      sycl::ext::intel::experimental::property::usm::buffer_location(0));
 #else
-  // allocate the shared memory as device memory allocation is not supported
-  // when targeting an FPGA family/part
-  T* in_ptr = malloc_shared<T>(size, q);
-  T* out_ptr = malloc_shared<T>(size, q);
+  // When targeting an FPGA family/part, use USM host or shared allocations
+  // since USM device allocations are not supported. Here we use USM shared
+  // allocation.
+  T *in_ptr = sycl::malloc_host<T>(
+      size, q,
+      sycl::ext::intel::experimental::property::usm::buffer_location(1));
+  T *out_ptr = sycl::malloc_host<T>(
+      size, q,
+      sycl::ext::intel::experimental::property::usm::buffer_location(1));
```

Comment:
The comment says "shared" so either the comment or the allocation function needs to be changed.
```suggestion
  T *in_ptr = sycl::malloc_shared<T>(
      size, q,
      sycl::ext::intel::experimental::property::usm::buffer_location(1));
  T *out_ptr = sycl::malloc_shared<T>(
      size, q,
      sycl::ext::intel::experimental::property::usm::buffer_location(1));
```

Reply:
Actually it doesn't matter which of the 2 is used. Neither of them require explicit copying around of data.